subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/NestedBeanAnnotationHandler.java

```diff
@@ -34,7 +41,35 @@ public boolean isPropertyRelevant() {
         return true;
     }
 
+    @Override
+    @SuppressWarnings("unchecked")
+    public void validatePropertyMetadata(PropertyMetadata propertyMetadata, TypeValidationContext validationContext) {
+        if (Map.class.isAssignableFrom(propertyMetadata.getDeclaredType().getRawType())) {
+            Class<?> keyType = extractNestedType((TypeToken<Map<?, ?>>) propertyMetadata.getDeclaredType(), Map.class, 0);
+            validateKeyType(propertyMetadata, validationContext, keyType);
+        }
+    }
+
     @Override
     public void visitPropertyValue(String propertyName, PropertyValue value, PropertyMetadata propertyMetadata, PropertyVisitor visitor) {
     }
+
+    private static <T> Class<?> extractNestedType(TypeToken<T> beanType, Class<? super T> parameterizedSuperClass, int typeParameterIndex) {
+        ParameterizedType type = (ParameterizedType) beanType.getSupertype(parameterizedSuperClass).getType();
+        return TypeToken.of(type.getActualTypeArguments()[typeParameterIndex]).getRawType();
+    }
+
+    private static void validateKeyType(PropertyMetadata propertyMetadata, TypeValidationContext validationContext, Class<?> keyType) {
+        if (!keyType.isAssignableFrom(String.class)) {
```

Comment:
:x: We should check for equality here instead, as a `Map` with `Object` keys would not hit this condition (since `Object.isAssignableFrom(String)`).